File: src/api/resources/sessions.ts
/**
 * 会话 API(mock ↔ http 双轨)。
 *
 * http 模式:GET /api/sessions?dept&method&nas&vlan&auth&q
 *   → SessionRowOut[] → SessionRow[](类型不变)
 *
 * disconnect:POST /api/sessions/disconnect
 */
import type { SessionRow } from '../../data/sessions';
import {
  SESSION_FILTER_OPTIONS,
  SESSION_ROWS,
} from '../../data/sessions';
import { MODE } from '../config';
import { downloadFile, fetchApi, fetchItems } from '../http';

export { SESSION_FILTER_OPTIONS, SESSION_ROWS };
export type { SessionRow };

export async function exportSessionsCsv(filters?: Record<string, string>): Promise<void> {
  if (MODE !== 'http') throw new Error('mock 模式不支持导出');
  const params = new URLSearchParams();
  if (filters) {
    for (const [k, v] of Object.entries(filters)) {
      if (!v || v === '全部' || v === '全部部门' || v === '全部设备') continue;
      if (k === 'method') params.set(k, _methodParam(v));
      else if (k === 'vlan') params.set(k, v.split(' ')[0] ?? v);
      else params.set(k, v);
    }
  }
  const qs = params.toString();
  await downloadFile(`/api/sessions/export.csv${qs ? '?' + qs : ''}`, 'sessions.csv');
}

// ── mock impl ────────────────────────────────────
async function mockFetch(): Promise<SessionRow[]> {
  return [...SESSION_ROWS];
}

function mockDisconnect(ids: string[]): Promise<{ disconnected: number; failed: { id: string; reason: string }[] }> {
  return Promise.resolve({ disconnected: ids.length, failed: [] });
}

// ── http impl ────────────────────────────────────
function _method(nasporttype: string): '有线' | 'WiFi' {
  const t = (nasporttype ?? '').toLowerCase();
  return t.includes('wireless') || t.includes('wi') ? 'WiFi' : '有线';
}

function _duration(duration_s: number): string {
  if (!duration_s || duration_s <= 0) return '0m';
  const h = Math.floor(duration_s / 3600);
  const m = Math.floor((duration_s % 3600) / 60);
  return h > 0 ? `${h}h ${m}m` : `${m}m`;
}

function mapSession(raw: any): SessionRow {
  return {
    session: raw.acct_unique_id ?? '',
    user: raw.username ?? '',
    name: raw.name ?? '',
    dept: raw.dept ?? '',
    mac: raw.mac ?? '',
    method: _method(raw.method),
    nas: raw.nas_name ?? '',
    nasSub: raw.nas_area ?? '',
    nasIp: raw.nas_ip ?? '',
    nasPort: raw.nas_port ?? '',
    called: raw.called ?? '',
    ip: raw.ip ?? '',
    vlan: raw.vlan ?? '',
    vlanLabel: raw.vlan_label ?? '',
    auth: raw.auth_method ?? '',
    duration: _duration(raw.duration_s),
    status: raw.status === 'online' ? '在线' : '待重认证',
    filterId: raw.filter_id ?? '',
    timeout: raw.session_timeout ?? '',
    start: raw.start ?? '',
  };
}

function _methodParam(label: string): string {
  if (label.includes('有线')) return '有线';
  if (label.toLowerCase().includes('wifi')) return 'WiFi';
  return label;
}

async function httpFetch(filters?: Record<string, string>): Promise<SessionRow[]> {
  const params = new URLSearchParams();
  if (filters) {
    for (const [k, v] of Object.entries(filters)) {
      if (!v || v === '全部' || v === '全部部门' || v === '全部设备') continue;
      if (k === 'method') params.set(k, _methodParam(v));
      else if (k === 'vlan') params.set(k, v.split(' ')[0] ?? v);
      else params.set(k, v);
    }
  }
  const qs = params.toString();
  const { items } = await fetchItems(`/api/sessions${qs ? '?' + qs : ''}`);
  return items.map(mapSession);
}

async function httpDisconnect(ids: string[]): Promise<{ disconnected: number; failed: { id: string; reason: string }[] }> {
  const body: any = await fetchApi('/api/sessions/disconnect', {
    method: 'POST',
    body: JSON.stringify({ session_ids: ids, confirm: true }),
  });
  return { disconnected: body.disconnected, failed: body.failed ?? [] };
}

// ── public API ───────────────────────────────────
export async function fetchSessions(filters?: Record<string, string>): Promise<SessionRow[]> {
  return MODE === 'http' ? httpFetch(filters) : mockFetch();
}

export function disconnectSessions(ids: string[]): Promise<{ disconnected: number; failed: { id: string; reason: string }[] }> {
  return MODE === 'http' ? httpDisconnect(ids) : mockDisconnect(ids);
}

File: backend/src/openredius/api/sessions.py
"""Online session endpoints (docs/03「在线会话」)."""

from __future__ import annotations

import asyncio
import csv
import io

import anyio
from fastapi import APIRouter, Depends, Query
from fastapi.responses import StreamingResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from openredius.core.config import Settings
from openredius.core.db import get_db
from openredius.core.deps import get_app_settings, require_role
from openredius.core.errors import ApiError
from openredius.core.listing import PageParams
from openredius.models import AdminRole, AdminUser
from openredius.radius.coa import CoaOutcome, disconnect_session
from openredius.radius.tables import radius_table
from openredius.schemas.sessions import (
    DisconnectFailure,
    DisconnectRequest,
    DisconnectResult,
    SessionDetailOut,
)
from openredius.services import audit
from openredius.services import sessions as svc

router = APIRouter()

_READ_ROLES = (AdminRole.ADMIN, AdminRole.OPERATOR, AdminRole.AUDITOR)
_WRITE_ROLES = (AdminRole.ADMIN, AdminRole.OPERATOR)
_DISCONNECT_CONCURRENCY = 8
_CLOSE_POLL_INTERVAL_S = 0.5


def _filters(
    dept: str | None = Query(None),
    method: str | None = Query(None),
    nas: str | None = Query(None),
    vlan: str | None = Query(None),
    auth: str | None = Query(None),
    q: str | None = Query(None),
) -> svc.SessionFilters:
    return svc.SessionFilters(dept=dept, method=method, nas=nas, vlan=vlan, auth=auth, q=q)


@router.get("")
async def list_sessions(
    db: AsyncSession = Depends(get_db),
    _admin: AdminUser = Depends(require_role(*_READ_ROLES)),
    filters: svc.SessionFilters = Depends(_filters),
    page: int = Query(1, ge=1),
    size: int = Query(50, ge=1, le=500),
    sort: str | None = Query(None),
) -> dict:
    return await svc.list_sessions(db, filters, PageParams(page, size), sort)


@router.get("/export.csv")
async def export_sessions_csv(
    db: AsyncSession = Depends(get_db),
    admin: AdminUser = Depends(require_role(*_READ_ROLES)),
    filters: svc.SessionFilters = Depends(_filters),
) -> StreamingResponse:
    rows = await svc.sessions_for_csv(db, filters)
    buf = io.StringIO()
    writer = csv.writer(buf)
    writer.writerow(_CSV_HEADER)
    for r in rows:
        data = r.model_dump()
        writer.writerow([data[h] for h in _CSV_HEADER])
    await audit.record_audit(
        db, actor=admin.username, action="session.export_csv", detail={"count": len(rows)}
    )
    await db.commit()
    content = buf.getvalue()
    return StreamingResponse(
        iter([content]),
        media_type="text/csv",
        headers={"Content-Disposition": "attachment; filename=sessions.csv"},
    )


_CSV_HEADER = [
    "acct_unique_id",
    "acct_session_id",
    "username",
    "name",
    "dept",
    "mac",
    "method",
    "nas_name",
    "nas_ip",
    "nas_port",
    "called",
    "ip",
    "vlan",
    "auth_method",
    "duration_s",
    "status",
    "filter_id",
    "session_timeout",
    "start",
    "bytes_in",
    "bytes_out",
]


@router.get("/{acct_unique_id}")
async def get_session(
    acct_unique_id: str,
    db: AsyncSession = Depends(get_db),
    _admin: AdminUser = Depends(require_role(*_READ_ROLES)),
) -> SessionDetailOut:
    detail = await svc.get_session_detail(db, acct_unique_id)
    if detail is None:
        raise ApiError("session_not_found", f"session {acct_unique_id} not found", 404)
    row, attributes = detail
    return SessionDetailOut(**row.model_dump(), attributes=attributes)


@router.post("/disconnect")
async def disconnect_sessions(
    body: DisconnectRequest,
    db: AsyncSession = Depends(get_db),
    admin: AdminUser = Depends(require_role(*_WRITE_ROLES)),
    settings: Settings = Depends(get_app_settings),
) -> DisconnectResult:
    if not body.confirm:
        raise ApiError("confirm_required", "disconnect requires confirm=true", 422)
    if not body.session_ids:
        raise ApiError("empty_selection", "session_ids must not be empty", 422)
    ids = list(dict.fromkeys(body.session_ids))
    rows = await svc.fetch_sessions_by_unique_ids(db, ids)
    found = {r.acct.acctuniqueid: r for r in rows}
    limiter = anyio.CapacityLimiter(_DISCONNECT_CONCURRENCY)

    async def send_coa(unique_id: str) -> tuple[str, CoaOutcome]:
        joined = found[unique_id]
        acct, nas = joined.acct, joined.nas
        if nas is None:
            return unique_id, CoaOutcome(status="error", error_cause="unknown-nas")
        async with limiter:
            outcome = await disconnect_session(
                host=nas.nasname,
                port=settings.radius_coa_port,
                secret=nas.secret_enc,
                username=acct.username,
                acct_session_id=acct.acctsessionid,
                calling_station_id=acct.callingstationid or None,
                timeout_s=settings.radius_coa_timeout,
            )
        return unique_id, outcome

    # Network fan-out first (no DB); DB polling/closing then runs serially on
    # this single session (AsyncSession is not concurrency-safe, and the SQLite
    # StaticPool test setup has exactly one connection).
    outcomes = dict(await asyncio.gather(*(send_coa(uid) for uid in ids if uid in found)))

    failed: list[DisconnectFailure] = []
    disconnected = 0
    for unique_id in ids:
        if unique_id not in found:
            failed.append(DisconnectFailure(id=unique_id, reason="not-found-or-not-active"))
            await audit.record_audit(
                db,
                actor=admin.username,
                action="session.disconnect",
                detail={"acct_unique_id": unique_id, "result": "not-found-or-not-active"},
            )
            continue
        outcome = outcomes[unique_id]
        if outcome.status == "ack":
            disconnected += 1
            await _await_stop_or_close(db, unique_id, settings.radius_coa_close_poll_s)
        else:
            failed.append(
                DisconnectFailure(id=unique_id, reason=outcome.error_cause or outcome.status)
            )
        await audit.record_audit(
            db,
            actor=admin.username,
            action="session.disconnect",
            detail={
                "acct_unique_id": unique_id,
                "result": outcome.status,
                "error_cause": outcome.error_cause,
            },
        )
    await db.commit()
    return DisconnectResult(disconnected=disconnected, failed=failed)


async def _await_stop_or_close(db: AsyncSession, unique_id: str, poll_s: float) -> None:
    """Poll radacct for the NAS-side stop; fall back to backend close (docs/04)."""
    loop = asyncio.get_running_loop()
    deadline = loop.time() + poll_s
    while True:
        radacct = radius_table(db.get_bind().dialect.name, "radacct")
        stop = (
            await db.execute(
                select(radacct.c.acctstoptime).where(radacct.c.acctuniqueid == unique_id)
            )
        ).scalar_one_or_none()
        if stop is not None:
            return
        if loop.time() >= deadline:
            await svc.close_session_fallback(db, unique_id)
            return
        await asyncio.sleep(_CLOSE_POLL_INTERVAL_S)
